util/src/network/host.rs

```diff
@@ -642,26 +642,25 @@ impl<Message> Host<Message> where Message: Send + Sync + Clone {
 
 		// turn a handshake into a session
 		let mut sessions = self.sessions.write().unwrap();
-		let mut h = handshakes.remove(token).unwrap();
-		// wait for other threads to stop using it
-		{
-			while Arc::get_mut(&mut h).is_none() {
-				h.lock().ok();
-			}
+		let mut h = handshakes.get_mut(token).unwrap().lock().unwrap();
+		if h.expired {
+			return;
 		}
-		let h = Arc::try_unwrap(h).ok().unwrap().into_inner().unwrap();
 		let originated = h.originated;
-		let mut session = match Session::new(h, &self.info.read().unwrap()) {
+		let mut session = match Session::new(&mut h, &self.info.read().unwrap()) {
```

Comment:
It is used later to deregister the handshake connection. This happens asynchronously in another event loop iteration.